### PR TITLE
Remove never_type feature requirement for exhaustive patterns

### LINKED ISSUE #51221: ICE when using exhaustive_patterns without never_type
```rust
~$ cat a.rs 
#![feature(exhaustive_patterns)]
fn main() {
    enum Void {}
    let a: Option<Void> = None;
    let None = a;
}
~$ rustc +nightly a.rs
error: internal compiler error: librustc_mir/build/matches/mod.rs:225: match pairs [MatchPair { place: _1, pattern: Pattern { ty: std::option::Option<main::Void>, span: a.rs:5:9: 5:13, kind: Variant { adt_def: std::option::Option, substs: [main::Void], variant_index: 0, subpatterns: [] } }, slice_len_checked: false }] remaining after simplifying irrefutable pattern
 --> a.rs:5:9
  |
5 |     let None = a;
  |         ^^^^

thread 'main' panicked at 'Box<Any>', librustc_errors/lib.rs:499:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: aborting due to previous error


note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.28.0-nightly (524ad9b9e 2018-05-29) running on x86_64-apple-darwin

👉	101
```

This compiles successfully with `#![feature(exhaustive_patterns,never_type)]`. It's possible that this regressed when the two features were separated. It's also possible that stabilizing never_type first will make this issue moot.

### PR DESCRIPTION
**Pull request by hellow554**
I **think** this resolves #51221
At least for me, it doesn't ICE anymore and all tests are still passing, so LGTM

### CODE REVIEW

File: src/test/run-pass/exhaustive_without_never_type.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#![feature(exhaustive_patterns)]
+// see issue #51221
```
Comment (Centril) [inline src/test/run-pass/exhaustive_without_never_type.rs]: This test can be `// check-pass`.

Also please move it to some appropriate location in `src/test/ui/` (e.g. with other tests for `exhaustive_patterns`) and also reference the issue number in the test name.

Comment (hellow554) [reply src/test/run-pass/exhaustive_without_never_type.rs]: @Centril  any idea for a place? Maybe https://github.com/rust-lang/rust/tree/master/src/test/ui/uninhabited ?

Comment (Centril) [reply src/test/run-pass/exhaustive_without_never_type.rs]: Yeah, that seems like a decent place.

**Comment by hellow554** [comment]
Done, waiting for <strike>travis</strike> azure

File: src/test/ui/uninhabited/exhaustive-wo-nevertype-issue-51221.rs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// check-pass
```
Comment (Centril) [inline src/test/ui/uninhabited/exhaustive-wo-nevertype-issue-51221.rs]: Seems the filename has a typo? "w" instead of "n"

Comment (hellow554) [reply src/test/ui/uninhabited/exhaustive-wo-nevertype-issue-51221.rs]: wo = without 😅
I was not sure how to name it. I didn't want only "issue-XXX" but also in this case describing that this is a general problem. 
Using exhaustive patterns should be possible regardless of the issue number.

Comment (Centril) [reply src/test/ui/uninhabited/exhaustive-wo-nevertype-issue-51221.rs]: Oh, I see; I thought you intended `no-nevertype` ^^

**Comment by hellow554** [comment]
@nikomatsakis do you have time for this, or should I pick another reviewer?

**Comment by hellow554** [comment]
r?

**Comment by hellow554** [comment]
@Dylan-DPC can you assign a new reviewer please?

**Comment by Dylan-DPC** [comment]
@hellow554 sure. can someone from @rust-lang/compiler review this?

**Comment by varkor** [comment]
As far as I can tell, when the `exhaustive_patterns` feature gate was added the `never_type` requirement was superseded, but was never removed, so this looks correct.

@bors r+ rollup

**Pull request merged**